src/h3/stream.rs

```diff
@@ -370,51 +332,130 @@ impl Stream {
 
                     _ => (),
                 }
-
-                self.state = State::FramePayloadLenLen;
             },
 
             _ => return Err(Error::UnexpectedFrame),
         }
 
         self.frame_type = Some(ty);
 
+        self.state_transition(State::FramePayloadLen, 1);
+
         Ok(())
     }
 
-    pub fn parse_frame(&mut self) -> Result<()> {
-        // Parse the whole frame payload but only if there is enough data in
-        // the stream buffer. stream.buf_bytes() returns an error if we don't
-        // have enough.
-        if self.frame_type.is_none() {
-            return Err(Error::InternalError);
+    /// Sets the frame's payload length and transitions to the next state.
+    pub fn set_frame_payload_len(&mut self, len: u64) -> Result<()> {
+        assert_eq!(self.state, State::FramePayloadLen);
+
+        // Only expect frames on Control, Request and Push streams.
+        if self.ty == Some(Type::Control) ||
+            self.ty == Some(Type::Request) ||
+            self.ty == Some(Type::Push)
+        {
+            self.state_transition(State::FramePayload, len as usize);
+
+            return Ok(());
         }
 
+        Err(Error::InternalError)
+    }
+
+    /// Tries to fill the state buffer by reading data from the corresponding
+    /// transport stream.
+    ///
+    /// When not enugh data can be read to complete the state, this returns
+    /// `Error::Done`.
+    pub fn try_fill_buffer(
+        &mut self, conn: &mut crate::Connection,
+    ) -> Result<()> {
+        let buf = &mut self.state_buf[self.state_off..self.state_len];
+
+        let (read, _) = conn.stream_recv(self.id, buf)?;
```

Comment:
passing `conn` lets us avoid one of the copies of the read data right? nice.

Reply:
Right. Basically we can just read the bytes directly into the stream buffer, instead of using an intermediate buffer like before.